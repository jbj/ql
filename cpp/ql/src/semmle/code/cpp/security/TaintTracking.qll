import semmle.code.cpp.ir.dataflow.DefaultTaintTracking
