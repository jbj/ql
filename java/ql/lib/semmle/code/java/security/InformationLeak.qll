/** Provides classes to reason about System Information Leak vulnerabilities. */

import java
import semmle.code.java.dataflow.DataFlow
private import semmle.code.java.dataflow.ExternalFlow
import semmle.code.java.security.XSS

/**
 * A sink that represent a method that outputs data to an HTTP response. Extend
 * this class to add more sinks that should be considered information leak
 * points by every query.
 */
abstract class AbstractInformationLeakSink extends DataFlow::Node { }

/**
 * A sink that represent a method that outputs data to an HTTP response. To add
 * more sinks, extend `AbstractInformationLeakSink` rather than this class.
 */
final class InformationLeakSink extends DataFlow::Node instanceof InformationLeakDiffInformed<xssNotDiffInformed/0>::InformationLeakSink
{ }

/** A default sink representing methods outputing data to an HTTP response. */
private class DefaultInformationLeakSink extends AbstractInformationLeakSink {
  DefaultInformationLeakSink() { sinkNode(this, "information-leak") }
}

/**
 * A module for finding XSS sinks faster in a diff-informed query. The
 * `querySource` parameter should be the `querySource` predicate for the
 * overall data-flow configuration of the query. In full generality,
 * `querySource` should hold for at least all the nodes that may be emitted in
 * an alert message together with a sink found by this module.
 */
module InformationLeakDiffInformed<xssNullaryPredicate/0 hasSourceInDiffRange> {
  final private class Node = DataFlow::Node;

  /**
   * A diff-informed replacement for the top-level `InformationLeakSink`,
   * omitting for efficiency some sinks that would never be reported by a
   * diff-informed query.
   */
  final class InformationLeakSink extends Node {
    InformationLeakSink() {
      this instanceof AbstractInformationLeakSink
      or
      this instanceof XssDiffInformed<hasSourceInDiffRange/0>::XssSink
    }
  }
}
