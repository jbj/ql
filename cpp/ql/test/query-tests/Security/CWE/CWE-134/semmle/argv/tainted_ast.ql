import semmle.code.cpp.security.TaintTrackingImpl

from Expr source, Element tainted
where tainted(source, tainted)
select source, tainted