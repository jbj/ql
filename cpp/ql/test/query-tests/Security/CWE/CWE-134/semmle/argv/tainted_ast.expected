| argvLocal.c:63:9:63:12 | argv | argvLocal.c:63:9:63:12 | argv |
| argvLocal.c:63:9:63:12 | argv | argvLocal.c:63:9:63:15 | access to array |
| argvLocal.c:64:15:64:18 | argv | argvLocal.c:64:15:64:18 | argv |
| argvLocal.c:64:15:64:18 | argv | argvLocal.c:64:15:64:21 | access to array |
| argvLocal.c:67:14:67:17 | argv | argvLocal.c:67:14:67:17 | argv |
| argvLocal.c:67:14:67:17 | argv | argvLocal.c:67:14:67:20 | access to array |
| argvLocal.c:95:9:95:12 | argv | argvLocal.c:6:24:6:29 | format |
| argvLocal.c:95:9:95:12 | argv | argvLocal.c:95:9:95:12 | argv |
| argvLocal.c:95:9:95:12 | argv | argvLocal.c:95:9:95:15 | (const char *)... |
| argvLocal.c:95:9:95:12 | argv | argvLocal.c:95:9:95:15 | access to array |
| argvLocal.c:96:15:96:18 | argv | argvLocal.c:6:24:6:29 | format |
| argvLocal.c:96:15:96:18 | argv | argvLocal.c:9:25:9:31 | correct |
| argvLocal.c:96:15:96:18 | argv | argvLocal.c:10:9:10:15 | (const char *)... |
| argvLocal.c:96:15:96:18 | argv | argvLocal.c:10:9:10:15 | correct |
| argvLocal.c:96:15:96:18 | argv | argvLocal.c:96:15:96:18 | argv |
| argvLocal.c:96:15:96:18 | argv | argvLocal.c:96:15:96:21 | access to array |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:6:24:6:29 | format |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:9:25:9:31 | correct |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:10:9:10:15 | (const char *)... |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:10:9:10:15 | correct |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:99:8:99:9 | i1 |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:100:2:100:3 | i1 |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:100:7:100:10 | argv |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:100:7:100:13 | access to array |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:101:9:101:10 | (const char *)... |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:101:9:101:10 | i1 |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:102:15:102:16 | i1 |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:143:8:143:9 | i7 |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:143:14:143:25 | ... , ... |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:143:24:143:25 | i1 |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:144:9:144:10 | (const char *)... |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:144:9:144:10 | i7 |
| argvLocal.c:100:7:100:10 | argv | argvLocal.c:145:15:145:16 | i7 |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:6:24:6:29 | format |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:9:25:9:31 | correct |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:10:9:10:15 | (const char *)... |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:10:9:10:15 | correct |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:105:9:105:10 | i2 |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:105:14:105:17 | argv |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:106:9:106:10 | i2 |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:106:9:106:13 | (const char *)... |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:106:9:106:13 | access to array |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:107:15:107:16 | i2 |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:107:15:107:19 | access to array |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:110:9:110:11 | (const char *)... |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:110:9:110:11 | * ... |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:110:10:110:11 | i2 |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:111:15:111:17 | * ... |
| argvLocal.c:105:14:105:17 | argv | argvLocal.c:111:16:111:17 | i2 |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:6:24:6:29 | format |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:7:20:7:21 | s1 |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:7:36:7:37 | s2 |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:9:25:9:31 | correct |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:10:9:10:15 | (const char *)... |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:10:9:10:15 | correct |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:114:7:114:8 | i3 |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:115:2:115:7 | call to memcpy |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:115:9:115:10 | i3 |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:115:13:115:16 | argv |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:115:13:115:19 | (const void *)... |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:115:13:115:19 | access to array |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:116:9:116:10 | i3 |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:117:15:117:16 | i3 |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:120:8:120:9 | i4 |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:120:13:120:14 | i3 |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:121:9:121:10 | (const char *)... |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:121:9:121:10 | i4 |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:122:15:122:16 | i4 |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:135:9:135:10 | i4 |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:135:9:135:12 | (const char *)... |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:135:9:135:12 | ... ++ |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:136:15:136:18 | -- ... |
| argvLocal.c:115:13:115:16 | argv | argvLocal.c:136:17:136:18 | i4 |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:6:24:6:29 | format |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:9:25:9:31 | correct |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:10:9:10:15 | (const char *)... |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:10:9:10:15 | correct |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:125:7:125:8 | i5 |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:126:2:126:3 | i5 |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:126:2:126:6 | access to array |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:126:10:126:13 | argv |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:126:10:126:16 | access to array |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:126:10:126:19 | access to array |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:127:9:127:10 | i5 |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:128:15:128:16 | i5 |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:131:9:131:10 | i5 |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:131:9:131:14 | (const char *)... |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:131:9:131:14 | ... + ... |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:132:15:132:16 | i5 |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:132:15:132:20 | ... + ... |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:139:9:139:26 | (const char *)... |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:139:9:139:26 | ... ? ... : ... |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:139:25:139:26 | i5 |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:140:15:140:32 | ... ? ... : ... |
| argvLocal.c:126:10:126:13 | argv | argvLocal.c:140:31:140:32 | i5 |
| argvLocal.c:139:9:139:12 | argv | argvLocal.c:139:9:139:12 | argv |
| argvLocal.c:139:9:139:12 | argv | argvLocal.c:139:9:139:15 | access to array |
| argvLocal.c:140:15:140:18 | argv | argvLocal.c:140:15:140:18 | argv |
| argvLocal.c:140:15:140:18 | argv | argvLocal.c:140:15:140:21 | access to array |
| argvLocal.c:143:14:143:17 | argv | argvLocal.c:143:14:143:17 | argv |
| argvLocal.c:143:14:143:17 | argv | argvLocal.c:143:14:143:20 | access to array |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:6:24:6:29 | format |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:9:25:9:31 | correct |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:10:9:10:15 | (const char *)... |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:10:9:10:15 | correct |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:148:8:148:9 | i8 |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:149:2:149:7 | * ... |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:149:4:149:6 | & ... |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:149:5:149:6 | i8 |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:149:11:149:14 | argv |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:149:11:149:17 | access to array |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:150:9:150:10 | (const char *)... |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:150:9:150:10 | i8 |
| argvLocal.c:149:11:149:14 | argv | argvLocal.c:151:15:151:16 | i8 |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:6:24:6:29 | format |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:7:20:7:21 | s1 |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:7:36:7:37 | s2 |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:9:25:9:31 | correct |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:10:9:10:15 | (const char *)... |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:10:9:10:15 | correct |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:155:8:155:9 | i9 |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:156:2:156:7 | call to memcpy |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:156:9:156:20 | ... ? ... : ... |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:156:13:156:16 | ++ ... |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:156:15:156:16 | i9 |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:156:23:156:26 | argv |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:156:23:156:29 | (const void *)... |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:156:23:156:29 | access to array |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:157:9:157:10 | (const char *)... |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:157:9:157:10 | i9 |
| argvLocal.c:156:23:156:26 | argv | argvLocal.c:158:15:158:16 | i9 |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:6:24:6:29 | format |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:7:20:7:21 | s1 |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:7:36:7:37 | s2 |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:9:25:9:31 | correct |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:10:9:10:15 | (const char *)... |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:10:9:10:15 | correct |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:162:8:162:10 | i91 |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:163:2:163:7 | call to memcpy |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:163:9:163:19 | ... ? ... : ... |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:163:17:163:19 | i91 |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:163:22:163:25 | argv |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:163:22:163:28 | access to array |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:163:22:163:32 | (const void *)... |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:163:22:163:32 | ... + ... |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:164:9:164:11 | (const char *)... |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:164:9:164:11 | i91 |
| argvLocal.c:163:22:163:25 | argv | argvLocal.c:165:15:165:17 | i91 |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:6:24:6:29 | format |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:9:25:9:31 | correct |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:10:9:10:15 | (const char *)... |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:10:9:10:15 | correct |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:168:6:168:8 | i10 |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:168:12:168:24 | (int)... |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:168:18:168:21 | argv |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:168:18:168:24 | access to array |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:169:9:169:20 | (char *)... |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:169:9:169:20 | (const char *)... |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:169:18:169:20 | i10 |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:170:15:170:26 | (char *)... |
| argvLocal.c:168:18:168:21 | argv | argvLocal.c:170:24:170:26 | i10 |
